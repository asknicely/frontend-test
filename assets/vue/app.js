/*/////  Light Vue Todos for Symfony2.0 /////
///////  Author: Kinson Lau ///////////////////////// */
import Vue from "vue";
import Todos from "./views/Todos";
import Todo from "./views/Todo";

/*
  VUE CONFIG
*/
Vue.config.devtools = true; // Remove for production
Vue.config.productionTip = false; // Update for production

/*
  ROOT COMPONENT
  METHODS
  - addToCart: Adds product via the product page and product tiles
  - updateQuantity: Updates an item in the cart
  - setFocus: Displays / hides the mobile search form and focuses
*/

new Vue({
  delimiters: ["${", "}"],
  el: "#app",
  components: {
    Todos: Todos,
    Todo: Todo
  }
});
