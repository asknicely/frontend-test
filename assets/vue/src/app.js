/* /////  Light Vue Todos - Symfony2.0 /////
///////  Author: Kinson Lau ///////////////////////// */

/*
  Import Vue components
*/
import Vue from 'vue'
import Todos from './views/Todos'

/*
  VUE CONFIG
*/
Vue.config.devtools = true // Remove for production
Vue.config.productionTip = false // Update for production

/*
  Init vue app
*/
new Vue({ // eslint-disable-line no-new
  delimiters: ['${', '}'],
  el: '#app',
  components: {
    Todos: Todos
  }
})
