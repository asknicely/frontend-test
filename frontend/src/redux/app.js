import React from 'react'

export const Context = React.createContext()

export const initial = {
    username: '',
    isLoggedIn: false,
    token: ''
}

export const reducer = (state, action) => {
    switch (action.type) {

        case 'DO_LOGIN':
            return { 
                ...state, 
                username: action.username,
                token: action.token,
                isLoggedIn: true
            }

        case 'DO_LOGOUT':
            return { 
                ...state, 
                username: '',
                token: '',
                isLoggedIn: false
            }            
        
        default:
            return {
                ...state,
            }
    }
}