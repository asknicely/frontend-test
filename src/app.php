<?php

use Silex\Application;
use Silex\Provider\SessionServiceProvider;
use Silex\Provider\TwigServiceProvider;
use Silex\Provider\UrlGeneratorServiceProvider;
use Silex\Provider\ValidatorServiceProvider;
use Silex\Provider\ServiceControllerServiceProvider;
use Silex\Provider\HttpFragmentServiceProvider;
use Silex\Provider\DoctrineServiceProvider;
use DerAlex\Silex\YamlConfigServiceProvider;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;

$app = new Application();
$app->register(new SessionServiceProvider());
$app->register(new UrlGeneratorServiceProvider());
$app->register(new ValidatorServiceProvider());
$app->register(new ServiceControllerServiceProvider());
$app->register(new TwigServiceProvider());
$app->register(new HttpFragmentServiceProvider());

$app->register(new YamlConfigServiceProvider(__DIR__.'/../config/config.yml'));
$app->register(new DoctrineServiceProvider, array(
    'db.options' => array(
        'driver'    => 'pdo_mysql',
        'host'      => $app['config']['database']['host'],
        'dbname'    => $app['config']['database']['dbname'],
        'user'      => $app['config']['database']['user'],
        'password'  => $app['config']['database']['password'],
        'charset'   => 'utf8',
    ),
));

$app->before(function (Request $request) use ($app) {
    if ($request->isMethod('OPTIONS')) {
        return new Response();
    }
});

$app->after(function (Request $request, Response $response) {
    $response->headers->set('Access-Control-Allow-Origin', $request->headers->get('origin'));
    $response->headers->set('Access-Control-Allow-Methods', 'GET, POST, PUT, PATCH, DELETE, OPTIONS');
    $response->headers->set('Access-Control-Allow-Credentials', 'true');
});

return $app;
