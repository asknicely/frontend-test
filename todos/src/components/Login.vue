<template>
    <div class="login-page">
        <div class="form">
            <form class="login-form" @submit.prevent="login()">
                <div class="input-group"
                     v-bind:class="{ error: errors.first('Username') }">
                    <input type="text" v-validate="'required'" placeholder="username" name="Username" v-model="formData.username"/>
                    <div class="error-message">{{ errors.first('Username') }} </div>
                </div>

                <div class="input-group"
                        v-bind:class="{ error: errors.first('Password') }">
                <input type="password" v-validate="'required'" placeholder="password"  name="Password" v-model="formData.password"/>
                <div class="error-message">{{ errors.first('Password') }}</div>
                </div>
                <button type="submit" value="Submit">Submit</button>
            </form>
        </div>
    </div>
</template>

<script>
    export default {
        name: 'Login',
        data() {
            return {
                formData: {
                    username: undefined,
                    password: undefined
                }
            };
        },
        methods:{
            async login(){

               const isValid = await this.$validator.validate();
               if(isValid){
                   console.log(this.formData.username);
                   const response = await this.$store.dispatch('user/login',
                       {username: this.formData.username, password: this.formData.password});
               }
            }
        }
    }
</script>

