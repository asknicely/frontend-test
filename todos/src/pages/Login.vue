<template>
    <div class="login-page">
        <div class="form">
            <form class="login-form" @submit.prevent="login()">
                <div class="input-group"
                     v-bind:class="{ error: errors.first('Username') }">
                    <input type="text" v-validate="'required'" placeholder="username" name="Username" v-model="formData.username"/>
                    <div class="error-message">{{ errors.first('Username') }} </div>
                </div>

                <div class="input-group"
                        v-bind:class="{ error: errors.first('Password') }">
                <input type="password" v-validate="'required'" placeholder="password"  name="Password" v-model="formData.password"/>
                <div class="error-message">{{ errors.first('Password') }}</div>
                </div>
                <div class="input-group"  v-bind:class="{ error: this.hasServerError }">
                <div class="error-message">User not found or your password is not correct.</div>
                </div>
                <button class='formbtn' type="submit" value="Submit">Submit</button>
            </form>
        </div>
    </div>
</template>

<script>
    export default {
        name: 'Login',
        data() {
            return {
                formData: {
                    username: undefined,
                    password: undefined
                },
                hasServerError:false
            };
        },
        methods:{
            async login(){

               const isValid = await this.$validator.validate();
               this.hasServerError=false;
               if(isValid){
                   try {
                       const response = await this.$store.dispatch('user/login',
                           {username: this.formData.username, password: this.formData.password});
                       this.$router.push({ name: 'todos' });
                   }
                   catch (e) {
                       console.log(e);
                       this.hasServerError=true
                   }
               }
            }
        }
    }
</script>

