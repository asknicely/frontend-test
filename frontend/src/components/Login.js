import React, { useState, useContext, useEffect } from 'react'
import axios from 'axios'
import { toast } from 'react-toastify'
import { Context } from '../redux/app'

const Login = (props) => {

	const defaultValue = {
		value: '',
		error: false
	}
	const [username, setUsername] = useState(defaultValue)
	const [password, setPassword] = useState(defaultValue)
	const { state, dispatch } = useContext(Context)

	const handleSubmit = async (e) => {

		e.preventDefault()
		var formError = 0

		if (!username.value) {
			formError++
			setUsername({...username, error: true})
		}

		if (!password.value) {
			formError++
			setPassword({...password, error: true})
		}

		if (formError) {
			toast.error('Please enter username and password')
		}
		else {

			// toast.info('New update is saved ...')
			var formData = new FormData()
			formData.append('username', username.value)
			formData.append('password', password.value)

			const url = `${process.env.REACT_APP_BACKEND_URL}/login`
			const response = await axios.post(url, formData)

			if (response.data.statusTxt !== 'ok') {
				toast.error('Login is unsuccessful ..')
				setUsername({...username, error: true})
				setPassword({...password, error: true})
			}
			else {

				localStorage.setItem('token', response.data.token)
				localStorage.setItem('username', username.value)
				localStorage.setItem('isLoggedIn', true)
				
				dispatch({
                    type: 'DO_LOGIN',
					username: username.value,
					token: response.data.token,
				})
				
				props.history.push('/todo')
				// toast.success('Welcome to TODOS App')
				
			}

		}

	}

	useEffect(() => { 
		document.title = 'Todos App - Login'
	}, [state])

	return (
		<div className="col-md-4 col-md-offset-4">
			<h1>Login</h1>
			<form className="form-horizontal login-form" method="POST" onSubmit={handleSubmit}>
				<div className="input-group">

					<span className="input-group-addon">
						<i className="glyphicon glyphicon-user"></i>
					</span>
					<input 
						type="text" 
						className={username.error ? 'form-control input-error' : 'form-control' }
						name="username" 
						placeholder="Username" 
						value={username.value ? username.value : ''}
						onChange={(e)=> setUsername({
							value: e.target.value,
							error: false
						})}
					/>

				</div>

				<div className="input-group">

					<span className="input-group-addon">
						<i className="glyphicon glyphicon-lock"></i>
					</span>
					<input 
						type="password" 
						className={password.error ? 'form-control input-error' : 'form-control' }
						name="password" 
						placeholder="Password" 
						value={password.value ? password.value : ''}
						onChange={(e)=> setPassword({
							value: e.target.value,
							error: false
						})}
					/>

				</div>
				<div className="form-group">
					<div className="col-sm-12 controls">
						<button type="submit" className="btn btn-primary pull-right">
							<i className="glyphicon glyphicon-log-in"></i> login
						</button>
					</div>
				</div>
			</form>
		</div>
	)
}

export default Login