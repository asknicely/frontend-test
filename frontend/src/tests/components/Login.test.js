import React, { useReducer } from 'react';
import { shallow, mount, render } from 'enzyme';
import Login from '../../components/Login';
import { initial, Context, reducer } from '../../redux/app'

const App = (props) => {
	const [state, dispatch] = useReducer(reducer, initial)
	return (
		<Context.Provider value={{ state, dispatch }}>
			<Login />
		</Context.Provider>
	);
}

test('should render Login correctly', () => {

	const wrapper = mount(<App />);
	expect(wrapper).toMatchSnapshot();
	
});


test('should give error on Login without filling user / pass', () => {

	const wrapper = mount(<App />);
	wrapper.find('form').simulate('submit', {
		preventDefault: () => { }
	});

	// make sure username has 'input-error' css class
	expect(wrapper.find('input').at(0).hasClass('input-error')).toBe(true);

	// make sure password has 'input-error' css class
	expect(wrapper.find('input').at(0).hasClass('input-error')).toBe(true);
	expect(wrapper).toMatchSnapshot();
	
});