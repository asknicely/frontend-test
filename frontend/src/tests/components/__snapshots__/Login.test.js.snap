// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should give error on Login without filling user / pass 1`] = `
<App>
  <Login>
    <div
      className="col-md-4 col-md-offset-4"
    >
      <h1>
        Login
      </h1>
      <form
        className="form-horizontal login-form"
        method="POST"
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="glyphicon glyphicon-user"
            />
          </span>
          <input
            className="form-control input-error"
            name="username"
            onChange={[Function]}
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="glyphicon glyphicon-lock"
            />
          </span>
          <input
            className="form-control input-error"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <div
            className="col-sm-12 controls"
          >
            <button
              className="btn btn-primary pull-right"
              type="submit"
            >
              <i
                className="glyphicon glyphicon-log-in"
              />
               login
            </button>
          </div>
        </div>
      </form>
    </div>
  </Login>
</App>
`;

exports[`should render Login correctly 1`] = `
<App>
  <Login>
    <div
      className="col-md-4 col-md-offset-4"
    >
      <h1>
        Login
      </h1>
      <form
        className="form-horizontal login-form"
        method="POST"
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="glyphicon glyphicon-user"
            />
          </span>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <div
          className="input-group"
        >
          <span
            className="input-group-addon"
          >
            <i
              className="glyphicon glyphicon-lock"
            />
          </span>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <div
            className="col-sm-12 controls"
          >
            <button
              className="btn btn-primary pull-right"
              type="submit"
            >
              <i
                className="glyphicon glyphicon-log-in"
              />
               login
            </button>
          </div>
        </div>
      </form>
    </div>
  </Login>
</App>
`;
